packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSource.ts

```diff
@@ -88,39 +113,62 @@ export const useGridDataSource = (
       const fetchParams = {
         ...gridGetRowsParamsSelector(apiRef),
         ...apiRef.current.unstable_applyPipeProcessors('getRowsParams', {}),
+        ...params,
       };
-
+      const startingIndex =
+        typeof fetchParams.start === 'string'
+          ? Math.max(filteredSortedRowIds.indexOf(fetchParams.start), 0)
+          : fetchParams.start;
       const cachedData = apiRef.current.unstable_dataSource.cache.get(fetchParams);
 
       if (cachedData !== undefined) {
         const rows = cachedData.rows;
-        apiRef.current.setRows(rows);
-        if (cachedData.rowCount) {
-          apiRef.current.setRowCount(cachedData.rowCount);
+        apiRef.current.setRowCount(cachedData.rowCount === undefined ? -1 : cachedData.rowCount);
+        if (isLazyLoaded) {
+          apiRef.current.unstable_replaceRows(startingIndex, rows);
+        } else {
+          apiRef.current.setRows(rows);
         }
+        apiRef.current.publishEvent('rowsFetched');
         return;
       }
 
+      // with lazy loading, only the initial load should show the loading overlay
+      const useLoadingIndicator = !isLazyLoaded || apiRef.current.getRowsCount() === 0;
```

Comment:
Possibly avoid using React hooks specific keyword `use`?
```suggestion
      const showLoadingIndicator = !isLazyLoaded || apiRef.current.getRowsCount() === 0;
```

Or maybe an inline condition would be more readable?
```tsx
if (!isLoading && (!isLazyLoaded || apiRef.current.getRowsCount() === 0)) {
```